packages/contracts/contracts/plugin/PluginSetupProcessor.sol

```diff
@@ -186,394 +246,388 @@ contract PluginSetupProcessor is DaoAuthorizable {
     }
 
     /// @notice Applies the permissions of a prepared installation to a DAO.
     /// @param _dao The address of the installing DAO.
-    /// @param _pluginSetup The address of the `PluginSetup` contract.
-    /// @param _pluginSetupRepo The repository storing the `PluginSetup` contracts of all versions of a plugin.
-    /// @param _plugin The address of the `Plugin` contract.
-    /// @param _permissions The list of multi-targeted permission operations to apply to the installing DAO.
-    function applyInstallation(
-        address _dao,
-        address _pluginSetup,
-        PluginRepo _pluginSetupRepo,
-        address _plugin,
-        PermissionLib.ItemMultiTarget[] calldata _permissions
-    ) external canApply(_dao, APPLY_INSTALLATION_PERMISSION_ID) {
-        // Check if the installation was applied already.
-        bytes32 appliedId = _getAppliedId(_dao, _plugin);
-        if (isInstallationApplied[appliedId]) {
-            revert SetupAlreadyApplied();
-        }
+    /// @param _params The struct containing the parameters for the `applyInstallation` function.
+    function applyInstallation(address _dao, ApplyInstallationParams calldata _params)
+        external
+        canApply(_dao, APPLY_INSTALLATION_PERMISSION_ID)
+    {
+        bytes32 pluginId = _getPluginId(_dao, _params.plugin);
 
-        bytes32 setupId = _getSetupId(_dao, _pluginSetup, address(_pluginSetupRepo), _plugin);
-        bytes32 storedPermissionsHash = installPermissionHashes[setupId];
+        PluginState storage pluginState = states[pluginId];
 
-        // Check if the installation was prepared.
-        if (storedPermissionsHash == bytes32(0)) {
-            revert SetupNotPrepared();
+        bytes32 setupId = _getSetupId(
+            _params.pluginSetupRef,
+            hashPermissions(_params.permissions),
+            _params.helpersHash,
+            bytes(""),
+            PreparationType.Install
+        );
+
+        // Allow calling `applyInstallation` only when
+        // plugin was uninstalled or never been installed before.
+        if (pluginState.currentSetupId != bytes32(0)) {
+            revert PluginAlreadyInstalled();
         }
 
-        // Check that the permissions match those announced in the preparation step.
-        if (storedPermissionsHash != _getPermissionsHash(_permissions)) {
-            revert PermissionsHashMismatch();
+        // If the plugin block number exceeds the setupId preparation block number,
+        // This means applyInstallation was already called on another setupId
+        // and all the rest setupIds should become idle or setupId is not prepared before.
+        if (pluginState.blockNumber >= pluginState.setupIds[setupId]) {
+            revert SetupNotApplicable(setupId);
         }
 
-        // Process the permission list.
-        DAO(payable(_dao)).bulkOnMultiTarget(_permissions);
+        bytes32 newSetupId = _getSetupId(
+            _params.pluginSetupRef,
+            bytes32(0),
```

Comment:
Perhaps a comment is needed explaining: why we don't include permission hashes again